### PR TITLE
crimson/os/seastore/segment_cleaner: correct available space calculation

### PR DESCRIPTION
**Pull request by xxhdx1985126**
Current available space calculation is wrong, it just counts the space occupied
by extents, deltas and other stuff are not taken into account.
    
Fixes: https://tracker.ceph.com/issues/53409
Signed-off-by: Xuehan Xu <xxhdx1985126@gmail.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by xxhdx1985126** [comment]
jenkins test make check

File: src/crimson/os/seastore/segment_cleaner.h
Comment (liu-chunmei) [inline src/crimson/os/seastore/segment_cleaner.h:202]: do you mean this branch is an error,  since the assert will always happen?  so give an error( ) message (not info) will be proper.

Comment (xxhdx1985126) [reply src/crimson/os/seastore/segment_cleaner.h:202]: yes, will correct it

File: src/crimson/os/seastore/segment_cleaner.h
Comment (athanatos) [inline src/crimson/os/seastore/segment_cleaner.h:1212]: Extra space

Comment (athanatos) [reply src/crimson/os/seastore/segment_cleaner.h:1212]: Also elsewhere in this file.

File: src/crimson/os/seastore/segment_cleaner.h
Comment (athanatos) [inline src/crimson/os/seastore/segment_cleaner.h:197]: Is this actually possible?

Comment (xxhdx1985126) [reply src/crimson/os/seastore/segment_cleaner.h:197]: Yes, I met this when testing this pr. It was because we do `SegmentCleaner::set_journal_head()` after the journal record is committed, at which point the segment might have been closed by later journal writes.

File: src/crimson/os/seastore/segment_cleaner.h
Comment (xxhdx1985126) [reply src/crimson/os/seastore/segment_cleaner.h:1212]: will correct them

**Comment by xxhdx1985126** [comment]
@athanatos @liu-chunmei Hi, I've modified the code as suggested, please take a look again:-)

File: src/crimson/os/seastore/segment_cleaner.h
Comment (cyx1231st) [inline src/crimson/os/seastore/segment_cleaner.h:1290]: The `avail_bytes` and `opened_segments` are at the `SegmentCleaner` level for all the managed devices, right?

But seems here (and below) they are assigned to be the stats by a specific device.

File: src/crimson/os/seastore/segment_cleaner.h
Comment (athanatos) [reply src/crimson/os/seastore/segment_cleaner.h:204]: Still looks like info()?

File: src/crimson/os/seastore/segment_cleaner.h
Comment (athanatos) [inline src/crimson/os/seastore/segment_cleaner.h:1344]: assert(stats.empty_segments > 0);

File: src/crimson/os/seastore/segment_cleaner.h
Comment (athanatos) [inline src/crimson/os/seastore/segment_cleaner.h:1286]: assert(stats.empty_segments > 0);

File: src/crimson/os/seastore/segment_cleaner.h
Comment (athanatos) [inline src/crimson/os/seastore/segment_cleaner.h:1225]: assert(stats.ios_blocking > 0);

File: src/crimson/os/seastore/segment_cleaner.h
Comment (xxhdx1985126) [reply src/crimson/os/seastore/segment_cleaner.h:632]: yes, will remove it

File: src/crimson/os/seastore/segment_cleaner.h
Comment (xxhdx1985126) [reply src/crimson/os/seastore/segment_cleaner.h:671]: yes

File: src/crimson/os/seastore/segment_cleaner.h
Comment (xxhdx1985126) [reply src/crimson/os/seastore/segment_cleaner.h:1290]: ah, yes

File: src/crimson/os/seastore/segment_cleaner.h
Comment (xxhdx1985126) [reply src/crimson/os/seastore/segment_cleaner.h:207]: Sorry, I modified an earlier "info". Just completed this change

**Comment by xxhdx1985126** [comment]
@athanatos @cyx1231st @liu-chunmei All done, please take a look again:-)

**Review by liu-chunmei — APPROVED** [review]
LGTM.

File: src/crimson/os/seastore/segment_cleaner.h
Comment (cyx1231st) [inline src/crimson/os/seastore/segment_cleaner.h:160]: Is it better named segment_opened now?

File: src/crimson/os/seastore/segment_cleaner.h
Comment (cyx1231st) [inline src/crimson/os/seastore/segment_cleaner.h:763]: This is still set by a specific device.

File: src/crimson/os/seastore/segment_cleaner.h
Comment (cyx1231st) [inline src/crimson/os/seastore/segment_cleaner.h:1302]: Can we remove both `stats.avail_bytes` and `stats.opened_segments` and always replace them by calling to `segment_info_set_t` interfaces directly?

Comment (xxhdx1985126) [reply src/crimson/os/seastore/segment_cleaner.h:1302]: The stats are for the seastar.metrics

Comment (cyx1231st) [reply src/crimson/os/seastore/segment_cleaner.h:1302]: Metrics don't require variables, they can also be registerred through lambdas:
https://github.com/ceph/ceph/blob/master/src/crimson/os/seastore/cache.cc#L468-L474

If we use stats.avail_bytes and stats.opened_segments, we still need them to be updated each time to get the most accurate values, right?

Comment (cyx1231st) [reply src/crimson/os/seastore/segment_cleaner.h:1302]: I think there are two considerations for a metric (whether it is a variable or through a lambda):
* It should be lightweight to be maintained/updated during I/O;
* It should be lightweight to be fetched because we may need to get a lot of metrics for each collection and we also don't want this to affect the test workload;

This needs careful implementations, so I'm fine to leave it to a follow-up PR.

Comment (xxhdx1985126) [reply src/crimson/os/seastore/segment_cleaner.h:1302]: > 

I think I can do this in this pr

**Comment by xxhdx1985126** [comment]
@athanatos @cyx1231st @liu-chunmei I've just modified the code as suggested, please take a look again

**Review by cyx1231st — APPROVED** [review]
LGTM except unresolved comments.

File: src/crimson/os/seastore/segment_cleaner.h
Comment (xxhdx1985126) [reply src/crimson/os/seastore/segment_cleaner.h]: done:-)

**Comment by athanatos** [comment]
retest this please

File: src/crimson/os/seastore/segment_cleaner.cc
Comment (athanatos) [inline src/crimson/os/seastore/segment_cleaner.cc:182]: stats.accumulated_blocked_ios

File: src/crimson/os/seastore/segment_cleaner.cc
Comment (athanatos) [inline src/crimson/os/seastore/segment_cleaner.cc:183]: "accumulated total number of ios that were blocked by gc"

File: src/crimson/os/seastore/segment_cleaner.h
Comment (athanatos) [inline src/crimson/os/seastore/segment_cleaner.h:180]: assert(segments.contains(segment));

File: src/crimson/os/seastore/segment_cleaner.h
Comment (athanatos) [inline src/crimson/os/seastore/segment_cleaner.h:184]: assert(sm_info->avail_bytes >= iter->second);
assert(avail_bytes >= iter->second);

File: src/crimson/os/seastore/segment_cleaner.h
Comment (athanatos) [inline src/crimson/os/seastore/segment_cleaner.h:218]: and these two.

**Review by athanatos — APPROVED** [review]
LGTM with the additional asserts.

I'm insistent on the asserts because it'll make the unit tests much more likely to detect any bugs introduced later.  Underflow bugs can be really difficult to find after the fact.

**Comment by xxhdx1985126** [comment]
> I'm insistent on the asserts because it'll make the unit tests much more likely to detect any bugs introduced later. Underflow bugs can be really difficult to find after the fact.

Will do:-)

**Comment by xxhdx1985126** [comment]
jenkins test make check

**Comment by xxhdx1985126** [comment]
> ../src/seastar/src/websocket/server.cc:122:51: error: no type named 'byte' in namespace 'CryptoPP'; did you mean simply 'byte'?
>    CryptoPP::StringSource(reinterpret_cast<const CryptoPP::byte*>(source.data()), source.size(),
>                                                  ^~~~~~~~~~~~~~
>                                                  byte
> /usr/include/cryptopp/config.h:203:23: note: 'byte' declared here
> typedef unsigned char byte;             // put in global namespace to avoid ambiguity with other byte typedefs

This doesn't seem to be caused by this PR, does anyone know why?

**Comment by athanatos** [comment]
Rebasing on master should fix it.

**Comment by xxhdx1985126** [comment]
> Rebasing on master should fix it, I think a fix for this merged yesterday.

Thanks:-)

**Comment by athanatos** [comment]
retest this please

**Comment by xxhdx1985126** [comment]
@athanatos Hi, sam, could you also take a final look at this PR:-) Thanks:-)

**Pull request merged**
